File: scripts/email-reservations/parse_huilo.py
import html as html_lib
import re
import unicodedata

MONTHS_ES = {
    "enero": 1,
    "febrero": 2,
    "marzo": 3,
    "abril": 4,
    "mayo": 5,
    "junio": 6,
    "julio": 7,
    "agosto": 8,
    "septiembre": 9,
    "setiembre": 9,
    "octubre": 10,
    "noviembre": 11,
    "diciembre": 12,
}

_LABELS = (
    r"Nombre",
    r"Alojamiento",
    r"N[uú]mero de confirmaci[oó]n",
    r"Pasajeros",
    r"Categor[ií]a de habitaci[oó]n",
    r"Tipo de habitaci[oó]n",
    r"Plan de alimentaci[oó]n",
    r"Traslados",
    r"Fechas",
    r"Tarifa final a pagar",
)


def html_to_text(html: str) -> str:
    """Tablas HTML de Huilo: <td>Label</td><td>valor</td> → Label: valor."""
    s = str(html or "")
    s = html_lib.unescape(s)
    s = s.replace("\xa0", " ").replace("\u200b", "")
    s = re.sub(r"(?i)<br\s*/?>", "\n", s)
    s = re.sub(r"(?i)</(p|div|tr|h[1-6]|li|table)>", "\n", s)
    s = re.sub(r"(?i)</t[dh]>\s*<t[dh][^>]*>", ": ", s)
    s = re.sub(r"(?i)</t[dh]>", "\n", s)
    s = re.sub(r"<[^>]+>", " ", s)
    s = s.replace("\r", "")
    s = re.sub(r"[ \t]+\n", "\n", s)
    s = re.sub(r"\n{3,}", "\n\n", s)
    s = re.sub(r"[ \t]{2,}", " ", s)
    labels = "|".join(_LABELS)
    s = re.sub(rf"(?im)(^|\n)\s*({labels})\s*\n\s*", r"\1\2: ", s)
    s = re.sub(r"(?im)(^|\n)\s*(In|Out)\s*\n\s*", r"\1\2: ", s)
    return s.strip()


def strip_html(html: str) -> str:
    return html_to_text(html)


def nombre_from_subject(subject: str):
    s = str(subject or "").strip()
    s = re.sub(r"^(RE|RV|FW|Fwd)\s*:\s*", "", s, flags=re.I)
    m = re.match(r"^Confirmaci[oó]n\s+(.+)$", s, re.I)
    if not m:
        return None
    name = m.group(1).strip()
    if re.search(r"fotograf|congreso|consulta|tarifas|fauna", name, re.I):
        return None
    return name or None


def _fold(s: str) -> str:
    return "".join(c for c in unicodedata.normalize("NFD", s) if unicodedata.category(c) != "Mn")


def parse_huilo_date(raw):
    if not raw:
        return None
    s = re.sub(r"\s+", " ", str(raw).strip().lower())
    s = re.sub(r"\s*[-–]\s*$", "", s).strip()
    m = re.search(r"(\d{1,2})[-\s/]+([a-záéíóú]+)[-\s/]+(\d{2,4})", s, re.I)
    if m:
        day = int(m.group(1))
        month = MONTHS_ES.get(_fold(m.group(2)))
        year = int(m.group(3))
        if year < 100:
            year += 2000
        if not month or not day:
            return None
        return f"{year:04d}-{month:02d}-{day:02d}"
    m = re.search(r"(\d{1,2})[/\-.](\d{1,2})[/\-.](\d{2,4})", s)
    if m:
        day, month, year = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if year < 100:
            year += 2000
        if month < 1 or month > 12 or day < 1 or day > 31:
            return None
        return f"{year:04d}-{month:02d}-{day:02d}"
    return None


def parse_passengers(raw):
    t = str(raw or "").lower()
    adults_m = re.search(r"(\d+)\s*adulto", t)
    child_m = re.search(r"\+\s*0?(\d+)\s*niñ", t)
    adults = int(adults_m.group(1)) if adults_m else 1
    children = int(child_m.group(1)) if child_m else 0
    return adults, children


def field(text, label):
    m = re.search(
        r"(?:^|\n)\s*" + label + r"\s*[:：]\s*(.+?)\s*(?=\n|$)",
        text,
        re.I | re.M,
    )
    return re.sub(r"\s+", " ", m.group(1)).strip() if m else None


def _from_text(text: str, subject: str, from_addr: str):
    looks = (
        "huilohuilo.com" in from_addr
        or re.search(r"huilo", subject, re.I)
        or re.search(r"Confirmaci[oó]n de reserva", text, re.I)
        or re.search(r"N[uú]mero de confirmaci[oó]n", text, re.I)
        or bool(nombre_from_subject(subject))
    )
    if not looks:
        return None
    if (
        not re.search(r"confirmaci[oó]n", subject, re.I)
        and not re.search(r"Confirmaci[oó]n de reserva", text, re.I)
        and not nombre_from_subject(subject)
    ):
        return None

    code = field(text, r"N[uú]mero de confirmaci[oó]n")
    if not code:
        m = re.search(r"N[uú]mero de confirmaci[oó]n\s*[:：]?\s*(.+)", text, re.I)
        code = m.group(1).strip() if m else None
    if code:
        nums = re.findall(r"\b\d{9}\b", code)
        if nums:
            code = " y ".join(dict.fromkeys(nums))
    if not code:
        nums = re.findall(r"\b\d{9}\b", text)
        if nums:
            code = " y ".join(dict.fromkeys(nums))
    nombre = field(text, r"Nombre") or nombre_from_subject(subject)
    if not code or not nombre:
        return None

    pasajeros_raw = field(text, r"Pasajeros") or ""
    adults, children = parse_passengers(pasajeros_raw)
    fechas_block = field(text, r"Fechas") or ""
    in_m = re.search(r"\bIn\s*[:：]?\s*(.+?)(?=\s*Out\b|$)", fechas_block, re.I)
    out_m = re.search(r"\bOut\s*[:：]?\s*(.+)$", fechas_block, re.I)
    in_line = re.search(r"\bIn\s*[:：]\s*([^\n]+)", text, re.I)
    out_line = re.search(r"\bOut\s*[:：]\s*([^\n]+)", text, re.I)
    check_in = parse_huilo_date(in_m.group(1) if in_m else None) or parse_huilo_date(
        in_line.group(1) if in_line else None
    )
    check_out = parse_huilo_date(out_m.group(1) if out_m else None) or parse_huilo_date(
        out_line.group(1) if out_line else None
    )

    tarifa_raw = field(text, r"Tarifa final a pagar") or ""
    if not tarifa_raw:
        m = re.search(r"Tarifa final a pagar\s*[:：]?\s*(.+)", text, re.I)
        tarifa_raw = m.group(1) if m else ""
    currency = "USD" if re.search(r"USD|US\$", tarifa_raw, re.I) else ("CLP" if re.search(r"CLP|\$", tarifa_raw) else "USD")
    amount_m = re.search(r"([\d]+(?:[.,]\d{2})?)", str(tarifa_raw).replace(".", ""))
    total_amount = float(amount_m.group(1).replace(",", ".")) if amount_m else 0.0

    if not check_in or not check_out:
        return None

    alojamiento = field(text, r"Alojamiento")
    room_category = field(text, r"Categor[ií]a de habitaci[oó]n")
    room_type = field(text, r"Tipo de habitaci[oó]n")
    meal_plan = field(text, r"Plan de alimentaci[oó]n")
    notes = " | ".join(
        x
        for x in [
            f"Alojamiento: {alojamiento}" if alojamiento else None,
            f"Habitación: {room_category}" if room_category else None,
            f"Tipo: {room_type}" if room_type else None,
            f"Plan: {meal_plan}" if meal_plan else None,
            f"Pasajeros: {pasajeros_raw}" if pasajeros_raw else None,
            f"Moneda: {currency}",
            "Origen: email Huilo (IMAP)",
        ]
        if x
    )
    return {
        "hotel_key": "huilo",
        "reservation_code": str(code).strip(),
        "client_name": nombre.strip(),
        "alojamiento": alojamiento,
        "room_category": room_category,
        "room_type": room_type,
        "meal_plan": meal_plan,
        "adults": adults,
        "children": children,
        "check_in": check_in,
        "check_out": check_out,
        "total_amount": total_amount,
        "currency": currency,
        "notes": notes,
    }


def parse_huilo_confirmation(mail: dict):
    from_addr = str(mail.get("from") or "").lower()
    subject = str(mail.get("subject") or "")
    text = mail.get("text") or ""
    html = mail.get("html") or ""
    candidates = []
    if text.strip():
        candidates.append(text)
    if html.strip():
        candidates.append(html_to_text(html))
    if not candidates:
        return None
    for body in candidates:
        parsed = _from_text(body, subject, from_addr)
        if parsed:
            return parsed
    return None

File: scripts/email-reservations/parse-huilo.js
/**
 * Parser de confirmaciones Huilo Huilo (mails de paz.galvez@huilohuilo.com).
 * Solo confirmaciones; modificaciones/cancelaciones van por WhatsApp.
 */

const MONTHS_ES = {
  enero: 1,
  febrero: 2,
  marzo: 3,
  abril: 4,
  mayo: 5,
  junio: 6,
  julio: 7,
  agosto: 8,
  septiembre: 9,
  setiembre: 9,
  octubre: 10,
  noviembre: 11,
  diciembre: 12,
};

function stripHtml(html) {
  return String(html || '')
    .replace(/<br\s*\/?>/gi, '\n')
    .replace(/<\/p>/gi, '\n')
    .replace(/<\/tr>/gi, '\n')
    .replace(/<\/div>/gi, '\n')
    .replace(/<[^>]+>/g, ' ')
    .replace(/&nbsp;/gi, ' ')
    .replace(/&amp;/gi, '&')
    .replace(/\r/g, '')
    .replace(/[ \t]+\n/g, '\n')
    .replace(/\n{3,}/g, '\n\n')
    .replace(/[ \t]{2,}/g, ' ')
    .trim();
}

/** @returns {string|null} YYYY-MM-DD */
function parseHuiloDate(raw) {
  if (!raw) return null;
  const s = String(raw).trim().toLowerCase().replace(/\s+/g, ' ');

  // 04-noviembre-2026 | 4 noviembre 2026
  let m = s.match(/^(\d{1,2})[-\s/]+([a-záéíóú]+)[-\s/]+(\d{2,4})$/i);
  if (m) {
    const day = parseInt(m[1], 10);
    const monName = m[2].normalize('NFD').replace(/\p{M}/gu, '');
    const month = MONTHS_ES[monName];
    let year = parseInt(m[3], 10);
    if (!month || !day) return null;
    if (year < 100) year += 2000;
    return `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
  }

  // 06-11-26 | 06/11/2026
  m = s.match(/^(\d{1,2})[\/\-.](\d{1,2})[\/\-.](\d{2,4})$/);
  if (m) {
    const day = parseInt(m[1], 10);
    const month = parseInt(m[2], 10);
    let year = parseInt(m[3], 10);
    if (year < 100) year += 2000;
    if (month < 1 || month > 12 || day < 1 || day > 31) return null;
    return `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
  }

  return null;
}

function parsePassengers(raw) {
  const t = String(raw || '').toLowerCase();
  const adultsM = t.match(/(\d+)\s*adulto/);
  const childM = t.match(/\+\s*0?(\d+)\s*niñ/);
  const adults = adultsM ? parseInt(adultsM[1], 10) : 1;
  const children = childM ? parseInt(childM[1], 10) : 0;
  return { adults, children };
}

function field(text, label) {
  const re = new RegExp(
    '(?:^|\\n)\\s*' + label + '\\s*[:：]\\s*(.+?)\\s*(?=\\n|$)',
    'im'
  );
  const m = text.match(re);
  return m ? m[1].replace(/\s+/g, ' ').trim() : null;
}

/**
 * @param {{ subject?: string, text?: string, html?: string, from?: string }} mail
 * @returns {null | {
 *   hotel_key: string,
 *   reservation_code: string,
 *   client_name: string,
 *   alojamiento: string|null,
 *   room_category: string|null,
 *   room_type: string|null,
 *   meal_plan: string|null,
 *   adults: number,
 *   children: number,
 *   check_in: string,
 *   check_out: string,
 *   total_amount: number,
 *   currency: string,
 *   notes: string
 * }}
 */
function parseHuiloConfirmation(mail) {
  const from = String(mail.from || '').toLowerCase();
  const subject = String(mail.subject || '');
  const text = mail.text || stripHtml(mail.html || '');

  const looksHuilo =
    from.includes('huilohuilo.com') ||
    /huilo/i.test(subject) ||
    /Confirmaci[oó]n de reserva/i.test(text) ||
    /N[uú]mero de confirmaci[oó]n/i.test(text);

  if (!looksHuilo) return null;
  if (!/confirmaci[oó]n/i.test(subject) && !/Confirmaci[oó]n de reserva/i.test(text)) {
    return null;
  }

  const code =
    field(text, 'N[uú]mero de confirmaci[oó]n') ||
    (text.match(/N[uú]mero de confirmaci[oó]n\s*[:：]?\s*(\d{6,})/i) || [])[1];
  const nombre = field(text, 'Nombre');
  if (!code || !nombre) return null;

  const pasajerosRaw = field(text, 'Pasajeros') || '';
  const { adults, children } = parsePassengers(pasajerosRaw);

  const fechasBlock = field(text, 'Fechas') || '';
  let checkIn =
    parseHuiloDate((fechasBlock.match(/\bIn\s*[:：]?\s*(.+?)(?=\s*Out\b|$)/i) || [])[1]) ||
    parseHuiloDate((text.match(/\bIn\s*[:：]\s*([^\n]+)/i) || [])[1]);
  let checkOut =
    parseHuiloDate((fechasBlock.match(/\bOut\s*[:：]?\s*(.+)$/i) || [])[1]) ||
    parseHuiloDate((text.match(/\bOut\s*[:：]\s*([^\n]+)/i) || [])[1]);

  const tarifaRaw =
    field(text, 'Tarifa final a pagar') ||
    (text.match(/Tarifa final a pagar\s*[:：]?\s*(.+)/i) || [])[1] ||
    '';
  const currency = /USD|US\$/i.test(tarifaRaw) ? 'USD' : /CLP|\$/i.test(tarifaRaw) ? 'CLP' : 'USD';
  const amountM = String(tarifaRaw).replace(/\./g, '').match(/([\d]+(?:[.,]\d{2})?)/);
  const total_amount = amountM ? parseFloat(amountM[1].replace(',', '.')) : 0;

  const alojamiento = field(text, 'Alojamiento');
  const room_category = field(text, 'Categor[ií]a de habitaci[oó]n');
  const room_type = field(text, 'Tipo de habitaci[oó]n');
  const meal_plan = field(text, 'Plan de alimentaci[oó]n');

  if (!checkIn || !checkOut) return null;

  const notesParts = [
    alojamiento && `Alojamiento: ${alojamiento}`,
    room_category && `Habitación: ${room_category}`,
    room_type && `Tipo: ${room_type}`,
    meal_plan && `Plan: ${meal_plan}`,
    pasajerosRaw && `Pasajeros: ${pasajerosRaw}`,
    `Moneda: ${currency}`,
    'Origen: email Huilo (IMAP)',
  ].filter(Boolean);

  return {
    hotel_key: 'huilo',
    reservation_code: String(code).trim(),
    client_name: nombre.trim(),
    alojamiento,
    room_category,
    room_type,
    meal_plan,
    adults,
    children,
    check_in: checkIn,
    check_out: checkOut,
    total_amount,
    currency,
    notes: notesParts.join(' | '),
  };
}

module.exports = {
  parseHuiloConfirmation,
  parseHuiloDate,
  stripHtml,
};
